sky/provision/paperspace/instance.py

```diff
@@ -0,0 +1,318 @@
+        logger.info(
+            f'Waiting for {num_restarted_instances}/{num_stopped_instances} '
+            'stopped instances to be restarted.')
+        time.sleep(POLL_INTERVAL)
+
+    exist_instances = _filter_instances(cluster_name_on_cloud,
+                                        status_filters=['ready'])
+    head_instance_id = _get_head_instance_id(exist_instances)
+    to_start_count = config.count - len(exist_instances)
+    if to_start_count < 0:
+        raise RuntimeError(
+            f'Cluster {cluster_name_on_cloud} already has '
+            f'{len(exist_instances)} nodes, but {config.count} are required.')
+    if to_start_count == 0:
+        if head_instance_id is None:
+            head_instance_id = list(exist_instances.keys())[0]
+            client.rename(
+                instance_id=head_instance_id,
+                name=f'{cluster_name_on_cloud}-head',
+            )
+        assert head_instance_id is not None, (
+            'head_instance_id should not be None')
+        logger.info(f'Cluster {cluster_name_on_cloud} already has '
+                    f'{len(exist_instances)} nodes, no need to start more.')
+        return common.ProvisionRecord(
+            provider_name='paperspace',
+            cluster_name=cluster_name_on_cloud,
+            region=region,
+            zone=None,
+            head_instance_id=head_instance_id,
+            resumed_instance_ids=list(newly_started_instances.keys()),
+            created_instance_ids=[],
+        )
+
+    created_instance_ids = []
+    for _ in range(to_start_count):
+        node_type = 'head' if head_instance_id is None else 'worker'
+        try:
+            instance_id = client.launch(
+                name=f'{cluster_name_on_cloud}-{node_type}',
+                instance_type=config.node_config['InstanceType'],
+                network_id=config.node_config['NetworkId'],
+                region=region,
+                disk_size=config.node_config['DiskSize'],
+            )['data']['id']
+        except Exception as e:  # pylint: disable=broad-except
+            logger.warning(f'run_instances error: {e}')
+            raise e
+        logger.info(f'Launched instance {instance_id}.')
+        created_instance_ids.append(instance_id)
+        if head_instance_id is None:
+            head_instance_id = instance_id
+
+    # Wait for instances to be ready.
+    for _ in range(MAX_POLLS_FOR_UP_OR_STOP):
+        instances = _filter_instances(cluster_name_on_cloud, ['ready'])
+        logger.info('Waiting for instances to be ready: '
+                    f'({len(instances)}/{config.count}).')
+        if len(instances) == config.count:
+            break
+
+        time.sleep(POLL_INTERVAL)
+    else:
+        # Failed to launch config.count of instances after max retries
+        msg = ('run_instances: Failed to create the'
+               'instances due to capacity issue.')
+        logger.warning(msg)
+        raise RuntimeError(msg)
```

Comment:
Just realized, it would be better to fail early if the capacity issue is detected, and we can then start failover to the other regions/clouds.
```suggestion
    for _ in range(MAX_POLLS_FOR_UP_OR_STOP):
        instances = _filter_instances(cluster_name_on_cloud, ['ready'])
        logger.info('Waiting for instances to be ready: '
                    f'({len(instances)}/{config.count}).')
        if len(instances) == config.count:
            break

        stopped_instances = _filter_instances(cluster_name_on_cloud, ['off'])
        # Paperspace will turn off the machine if there is a capacity issue
        if stopped_instances:
            msg = ('run_instances: Found a stopped instance during creation '
                   'likely due to capacity issue.')
            logger.warning(msg)
            raise RuntimeError(msg)
        time.sleep(POLL_INTERVAL)
    else:
        # Failed to launch config.count of instances after max retries
        msg = ('run_instances: Failed to create the'
               'instances due to capacity issue.')
        logger.warning(msg)
        raise RuntimeError(msg)
```